### PR TITLE
VA Profile: Implement JWT Auth for POST endpoint

### PR DESCRIPTION
**Pull request by kalbfled**
#728

When VA Profile makes a POST request to update our local cache, they should include an asymmetrically signed [JWT](https://en.wikipedia.org/wiki/JSON_Web_Token).  We will receive these requests from Profile via AWS API Gateway v2, which proxies the request to our lambda with a specific [event format](https://docs.aws.amazon.com/lambda/latest/dg/services-apigateway.html#apigateway-example-event) that I did not previously consider.  These updates:

- Update the lambda code to reflect the `event` it will actually receive
- Attempt to verify the JWT signature using an associated public key that is available in an env variable
- Update unit tests
- Include a minor README update 

I created a [PR in the infra repo](https://github.com/department-of-veterans-affairs/vanotify-infra/pull/392) to support changes required therein.

The changes in this PR also configure the local cache lambda to use the VA root pem to verify VA Profile's server when we make a PUT request to their endpoint.  Their server also authenticates our client (mTLS), but I haven't made any changes in this code to support that behavior.  I think their server will request and receive the default cert associated with our environment, and they can verify it with the VA room pem.

Note that these changes include an addition to requirements-for-test.txt.  Rebuild the test container before you run tests locally if you pull this branch or after this PR is merged.

Also note that these changes include several pem files for testing purposes.  One of these is a private key.  These pems are not used in any deployment environment.

### CODE REVIEW

**Review by maham360 — DISMISSED** [review]

**Comment by kalbfled** [comment]
I converted this PR to a draft because of the failing unit tests, which pass locally.  I deployed the associated infrastructure change to dev, verified the change in AWS Lambda console, and reran these remote tests.  The same tests failed again.

File: ci/docker-compose-test.yml [hunk 1]
```diff
@@ -18,6 +18,29 @@ services:
       - AWS_SECURITY_TOKEN=test
       - AWS_REGION=us-east-1
       - VA_PROFILE_DOMAIN=int.vaprofile.va.gov
+      - VA_PROFILE_PUBLIC_KEY=test
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply ci/docker-compose-test.yml:21]: It's wrong.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 2]
```diff
@@ -2,72 +2,103 @@
...
 
-if NOTIFY_ENVIRONMENT is None:
-    logging.error("Couldn't get the Notify environment.  This is necessary to retrieve the .pem file.")
-elif NOTIFY_ENVIRONMENT != "test":
-    PEM = get_pem()
+        # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ssm.html?highlight=get_parameter#SSM.Client.get_parameter
+        # TODO - "get_parameters" could be used to get the VA root pem and the db URI
+        # in one call.
+        response = ssm_client.get_parameter(
+            Name=f"/{NOTIFY_ENVIRONMENT}/notification-api/va-profile/pem",
```
Comment (jakehova) [inline lambda_functions/va_profile/va_profile_opt_in_out_lambda.py]: should this be va-root-pem?  I think I saw that in the infra repo

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 3]
```diff
@@ -184,42 +220,81 @@ def va_profile_opt_in_out_lambda_handler(event: dict, context, worker_id=None) -
...
         finally:
             put_request_body["bios"].append(put_record)
 
-    if len(put_request_body["bios"]) > 0 and PEM is not None and VA_PROFILE_DOMAIN is not None:
-        context = SSLContext()
+    if not requested_va_root_pem:
+        va_root_pem = get_va_root_pem()
+        requested_va_root_pem = True
 
+    if va_root_pem is not None and len(put_request_body["bios"]) > 0:
```
Comment (jakehova) [inline lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:237]: Can this just be: 

```
if va_root_pem is not None and put_request_body["bios"]:
```

**Review by jakehova — CHANGES_REQUESTED** [review]

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py (hunk 3 above)
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:237]: Can and should.  Will change.

**Review by jakehova — DISMISSED** [review]

**Comment by k-macmillan** [comment]
Need an expiration test in there. The agreed upon expiration of the JWT was 15 minutes.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 4]
```diff
@@ -2,72 +2,103 @@
...
-    """
 
-    ssm_client = boto3.client("ssm", region_name=os.getenv("AWS_REGION"))
+try:
+    # This is a public certificate in .pem format.  Use it to verify the signature
+    # of the JWT contained in POST requests from VA Profile.
+    va_profile_public_cert = load_pem_x509_certificate(VA_PROFILE_PUBLIC_KEY.encode()).public_key()
+except ValueError as e:
+    logger.exception(e)
+    logger.debug("VA_PROFILE_PUBLIC_KEY =\n", VA_PROFILE_PUBLIC_KEY)
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:54]:
```
>>> logger = logging.getLogger()
>>> logger.setLevel(logging.INFO)
>>> logger.info("Log this.\nAnd this.")
INFO:root:Log this.
And this.
```

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 5]
```diff
@@ -2,72 +2,103 @@
...
+        # TODO - "get_parameters" could be used to get the VA root pem and the db URI
+        # in one call.
+        response = ssm_client.get_parameter(
+            Name=f"/{NOTIFY_ENVIRONMENT}/notification-api/va-profile/pem",
+            WithDecryption=True
+        )
 
-    if PEM is None:
-        logging.error("Couldn't get the .pem chain from SSM.")
+        the_pem = response.get("Parameter", {}).get("Value")
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:92]: None is the default value.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (k-macmillan) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:54]: Wish I had an example so I knew what was done wrong in the ones that wouldn't put it on a newline. Oh well.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 6]
```diff
@@ -115,41 +150,42 @@ def va_profile_opt_in_out_lambda_handler(event: dict, context, worker_id=None) -
...
     """
 
-    global db_connection, pem
-    logging.debug(event)
+    logger.debug(event)
 
-    if "txAuditId" not in event or "bios" not in event or not isinstance(event["bios"], list):
-        return { "statusCode": 400, "body": "A required top level attribute is missing from the request or has the wrong type." }
+    headers = event.get("headers", {})
+    if not jwt_is_valid(headers.get("Authorization", headers.get("authorization", ''))):
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:161]: The headers are user input, and I'm not sure if clients standardize capitalization.  This is my attempt to preclude an error.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 7]
```diff
@@ -184,42 +220,81 @@ def va_profile_opt_in_out_lambda_handler(event: dict, context, worker_id=None) -
...
+
+    if bearer.title() != "Bearer":
+        logger.debug("Malformed Authorization header value: ", auth_header_value)
+        return False
+
+    try:
+        # This returns the claims as a dictionary, but we aren't using them.  Require the
+        # Issued at Time (iat) claim to ensure the JWT varies with each request.  Otherwise,
+        # an attacker could replay the static Bearer value.
+        jwt.decode(token, va_profile_public_cert, algorithms=["RS256"], options={"require": ["iat"]})
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:294]: Added with additional unit tests.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 8]
```diff
@@ -2,72 +2,103 @@
...
-      https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/boto3.html?highlight=boto3.client#boto3.client
-    """
 
-    ssm_client = boto3.client("ssm", region_name=os.getenv("AWS_REGION"))
+try:
+    # This is a public certificate in .pem format.  Use it to verify the signature
+    # of the JWT contained in POST requests from VA Profile.
+    va_profile_public_cert = load_pem_x509_certificate(VA_PROFILE_PUBLIC_KEY.encode()).public_key()
+except ValueError as e:
+    logger.exception(e)
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:53]: This is referring to an "except" block.

File: lambda_functions/va_profile/va_profile_opt_in_out_lambda.py [hunk 9]
```diff
@@ -184,42 +220,81 @@ def va_profile_opt_in_out_lambda_handler(event: dict, context, worker_id=None) -
...
+    try:
+        # This returns the claims as a dictionary, but we aren't using them.  Require the
+        # Issued at Time (iat) claim to ensure the JWT varies with each request.  Otherwise,
+        # an attacker could replay the static Bearer value.
+        jwt.decode(token, va_profile_public_cert, algorithms=["RS256"], options={"require": ["iat"]})
+        return True
+    except jwt.exceptions.InvalidTokenError as e:
+        logger.exception(e)
+        logger.debug(auth_header_value)
+
```
[the remark this answers is not in the document]
Comment (kalbfled) [reply lambda_functions/va_profile/va_profile_opt_in_out_lambda.py:304]: "except" block

**Review by k-macmillan — APPROVED** [review]

**Review by EvanParish — APPROVED** [review]

**Pull request merged**